Backend/Remora.Discord.Caching.Redis/Services/RedisCacheProvider.cs

```diff
@@ -0,0 +1,113 @@
+//
+//  RedisCacheProvider.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Objects;
+using Remora.Results;
+
+namespace Remora.Discord.Caching.Services;
+
+/// <summary>
+/// Handles cache insert/evict operations for various types, using Redis as a backing-store.
+/// </summary>
+public class RedisCacheProvider : ICacheProvider
+{
+    private readonly CacheSettings _cacheSettings;
+    private readonly IDistributedCache _cache;
+    private readonly JsonSerializerOptions _jsonOptions;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RedisCacheProvider"/> class.
+    /// </summary>
+    /// <param name="settings">The settings for the cache.</param>
+    /// <param name="cache">The redis cache.</param>
+    /// <param name="jsonOptions">The JSON options.</param>
+    public RedisCacheProvider(IOptions<CacheSettings> settings, IDistributedCache cache, IOptionsMonitor<JsonSerializerOptions> jsonOptions)
+    {
+        _cacheSettings = settings.Value;
+        _cache = cache;
+        _jsonOptions = jsonOptions.Get("Discord");
+    }
+
+    /// <inheritdoc cref="ICacheProvider.CacheAsync{TInstance}"/>
+    public async ValueTask CacheAsync<TInstance>
+        (string key, TInstance instance, TimeSpan? absoluteExpiration = null, TimeSpan? slidingExpiration = null, CancellationToken ct = default)
+        where TInstance : class
+    {
+        if (absoluteExpiration == TimeSpan.Zero)
+        {
+            return;
+        }
+
+        var serialized = JsonSerializer.Serialize(instance, _jsonOptions);
```

Comment:
What happens if the type is not trivially serializable or is otherwise configured for JSON serialization here? This works fine for Discord entities, but in the case of more complex objects where the memory cache handles everything users might end up with unexpected behaviour.

Reply:
Did you have any particular example?

Reply:
In particular, I'm thinking of `RateLimitBucket`, which does not currently have well-defined serialization. Generally, this is a concern because the method signatures on `ICacheProvider` does not imply that the instance that is to be cached must be JSON serializable, and as such users might try to cache any arbitrary object - I certainly would. If we want this to be a requirement, it should be made much clearer, and should have some sort of additional check beyond just doing the serialization and hoping for the best.

Reply:
Serialization could of course be added for types we control, but that does not neccesarily translate to consumers of the library.

Reply:
Fair enough. Not sure how you'd even go about checking for such a thing, though.